File: models/Gpt-3.5-turbo/cpp/code/problems/graph_algorithms/Counting_Rooms.cpp

#include <iostream>
#include <vector>
using namespace std;

vector<vector<char>> map;
vector<vector<bool>> visited;

void dfs(int x, int y) {
    if (x < 0 || x >= map.size() || y < 0 || y >= map[0].size() || visited[x][y])
        return;

    visited[x][y] = true;

    if (map[x][y] == '#')
        return;

    dfs(x - 1, y);
    dfs(x + 1, y);
    dfs(x, y - 1);
    dfs(x, y + 1);
}

int main() {
    int n, m;
    cin >> n >> m;

    map.resize(n, vector<char>(m));
    visited.resize(n, vector<bool>(m, false));

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> map[i][j];
        }
    }

    int rooms = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            if (!visited[i][j] && map[i][j] == '.') {
                dfs(i, j);
                rooms++;
            }
        }
    }

    cout << rooms << endl;

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/graph_algorithms/Counting_Rooms.py

def dfs(i, j, n, m, map, visited):
    if i < 0 or i >= n or j < 0 or j >= m:
        return
    if visited[i][j]:
        return
    if map[i][j] == '#':
        return

    visited[i][j] = True

    # Recursively explore adjacent floor squares
    dfs(i+1, j, n, m, map, visited)
    dfs(i-1, j, n, m, map, visited)
    dfs(i, j+1, n, m, map, visited)
    dfs(i, j-1, n, m, map, visited)

def count_rooms(n, m, map):
    visited = [[False] * m for _ in range(n)]
    count = 0

    for i in range(n):
        for j in range(m):
            if not visited[i][j] and map[i][j] == '.':
                dfs(i, j, n, m, map, visited)
                count += 1

    return count

# Read input
n, m = map(int, input().split())
map = [input() for _ in range(n)]

# Count rooms
rooms = count_rooms(n, m, map)

# Print result
print(rooms)
